src/main/java/gregtech/common/metatileentities/storage/CraftingRecipeLogic.java

```diff
@@ -1,213 +1,451 @@
+    private IItemHandlerModifiable availableHandlers;
+    private final Hash.Strategy<ItemStack> strategy = ItemStackHashStrategy.builder()
+            .compareItem(true)
+            .compareMetadata(true)
+            .build();
+
+    /**
+     * Used to lookup a list of slots for a given stack
+     * filled by {@link CraftingRecipeLogic#refreshStackMap()}
+     **/
+    private final Map<ItemStack, Set<Integer>> stackLookupMap = new Object2ObjectOpenCustomHashMap<>(this.strategy);
+
+    /**
+     * List of items needed to complete the crafting recipe, filled by
+     * {@link CraftingRecipeLogic#getIngredientEquivalent(CraftingInputSlot)} )}
+     **/
+    private final Map<ItemStack, Integer> requiredItems = new Object2IntOpenCustomHashMap<>(
+            this.strategy);
+
+    private final Map<Integer, Object2BooleanMap<ItemStack>> replaceAttemptMap = new Int2ObjectArrayMap<>();
+    private final InventoryCrafting craftingMatrix;
     private final IInventory craftingResultInventory = new InventoryCraftResult();
-    private ItemStack oldResult = ItemStack.EMPTY;
     private final CachedRecipeData cachedRecipeData;
-    private final CraftingRecipeMemory recipeMemory;
-    private IRecipe cachedRecipe = null;
-    private int itemsCrafted = 0;
-    public static short ALL_INGREDIENTS_PRESENT = 511;
-    private short tintLocation = ALL_INGREDIENTS_PRESENT;
+    private final CraftingInputSlot[] inputSlots = new CraftingInputSlot[9];
 
-    public CraftingRecipeLogic(ICraftingStorage craftingStorage) {
-        this.world = craftingStorage.getWorld();
-        this.craftingGrid = craftingStorage.getCraftingGrid();
-        this.recipeMemory = craftingStorage.getRecipeMemory();
-        this.itemSources = new ItemSources(world);
-        this.cachedRecipeData = new CachedRecipeData(itemSources, null, inventoryCrafting);
-    }
-
-    public ItemSources getItemSourceList() {
-        return itemSources;
+    public CraftingRecipeLogic(World world, IItemHandlerModifiable handlers, IItemHandlerModifiable craftingMatrix) {
+        this.world = world;
+        this.availableHandlers = handlers;
+        this.craftingMatrix = wrapHandler(craftingMatrix);
+        this.cachedRecipeData = new CachedRecipeData();
     }
 
     public IInventory getCraftingResultInventory() {
         return craftingResultInventory;
     }
 
-    public int getItemsCraftedAmount() {
-        return itemsCrafted;
+    public InventoryCrafting getCraftingMatrix() {
+        return this.craftingMatrix;
     }
 
-    public void setItemsCraftedAmount(int itemsCrafted) {
-        this.itemsCrafted = itemsCrafted;
+    public void updateInventory(IItemHandlerModifiable handler) {
+        this.availableHandlers = handler;
     }
 
     public void clearCraftingGrid() {
         fillCraftingGrid(Collections.emptyMap());
     }
 
     public void fillCraftingGrid(Map<Integer, ItemStack> ingredients) {
```

Comment:
Does this need to be passed a Map? Can't it just be passed a ItemStack to fill the grid with? The only use of this method is to clear the grid by passing an empty map to fill with empty itemstacks.